topsites.py

```diff
@@ -0,0 +1,180 @@
+Base = declarative_base()
+Session = sessionmaker(bind=engine)
+session = Session()
+
+
+class Site(Base):
+    '''Define a sqlalchemy base object for a alexa top site.'''
+    __tablename__ = "topsites"
+
+    url = Column(String, primary_key=True)
+    priority = Column(Integer)
+    country_code = Column(String)
+    ranking = Column(Integer)
+
+    def __init__(self, url, priority, country_code, ranking):
+        self.url = url
+        self.priority = priority
+        self.country_code = country_code
+        self.ranking = ranking
+
+Base.metadata.create_all(engine)
+
+
+class TopSites:
+    '''Used to query top site info from alexa API to Site object'''
+    def __init__(self, access_key, secret_key):
+        self.access_key = access_key
+        self.secret_key = secret_key
+
+    def query(self, begin, begin_priority):
+        '''Query alexa with given range, save to Site'''
+        query_string = self.build_query(begin, 100)
+        toSign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+        signature = self.gen_sign(toSign, self.secret_key)
+        uri = "{base}{query}&{signature}".format(
+            base=ATS_AWS_BASE_URL,
+            query=query_string,
+            signature=urlencode({"Signature": signature}, "UTF-8"))
+
+        response = requests.get(uri)
+
+        if response.status_code == requests.codes.ok:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html  # nopep8
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                url = self.get_text(
+                    site.getElementsByTagName("aws:DataUrl")[0].childNodes)
+                rank = int(self.get_text(
+                    site.getElementsByTagName("aws:Rank")[0].childNodes))
+
+                priority = begin_priority
+                if rank > 100 and rank <= 1000:
+                    priority = priority + 1
+                elif rank > 1000 and rank <= 10000:
+                    priority = priority + 2
+
+                if url in topsites:
+                    site_row = topsites[url]
+                    # Update priority is new is lower
+                    if site_row.priority > priority:
+                        site_row.priority = priority
+                        site_row.country_code = self.country_code
+                        site_row.ranking = rank
+                else:
+                    site_row = Site(url, priority, self.country_code, rank)
+                    topsites[url] = site_row
+                    session.add(site_row)
+
+            session.commit()
+
+    def build_query(self, start, count):
+        '''Build query string for request with start ranking and count'''
+        now = time.time()
+        mlsec = repr(now).split('.')[1][:3]
+        nowtime = datetime.datetime.utcnow()
+        timestamp = nowtime.strftime(ATS_DATEFORMAT).format(mlsec)
```

Comment:
The sample code and sample query from amazon gives miliseconds. But seems work without that.
```
&Timestamp=2011-05-06T17%3A58%3A49.463Z
```